### ISSUE TITLE
ND_R2D visualization with glvis fails

### ISSUE THREAD

**Issue by Rickbude** [mfem/mfem#4403]
glvis seems not capable of visualizing the ND_R2D elements. To reproduce, run example 31:

``` 
ex31 -m ../data/star.mesh -o 2
```

and try to visualize with glvis, with the command suggested in the code:

```
glvis -m refined.mesh -g sol.gf
```

This gives as error:

``` 
Verification failed: (dim >= 1 && dim <= 3) is false:
 --> ND_FECollection requires 1 <= dim <= 3.
 ... in function: mfem::ND_FECollection::ND_FECollection(int, int, int, int)
 ... in file: /home/rick/tmp/mfem/fem/fe_coll.cpp:2708

Aborted (core dumped)
```

At first I thought that this was a glvis issue, but when I look at the trace with gdb, it seems like MFEM tries to load the wrong finite element space (ND instead of ND_R2D):

``` 
#0  __pthread_kill_implementation (threadid=<optimized out>, signo=signo@entry=6, no_tid=no_tid@entry=0) at pthread_kill.c:44
#1  0x00007ffff6094a73 in __pthread_kill_internal (signo=6, threadid=<optimized out>) at pthread_kill.c:78
#2  0x00007ffff6041176 in __GI_raise (sig=sig@entry=6) at ../sysdeps/posix/raise.c:26
#3  0x00007ffff6028917 in __GI_abort () at abort.c:79
#4  0x00007ffff73e32bd in mfem::mfem_error (msg=0x990cb0 "Verification failed: (dim >= 1 && dim <= 3) is false:\n --> ND_FECollection requires 1 <= dim <= 3.\n ... in function: mfem::ND_FECollection::ND_FECollection(int, int, int, int)\n ... in file: /home/rick"...)
    at /home/rick/tmp/mfem/general/error.cpp:184
#5  0x00007ffff7921367 in mfem::ND_FECollection::ND_FECollection (this=0x990800, p=2, dim=0, cb_type=1, ob_type=0) at /usr/include/c++/13/bits/basic_string.h:222
#6  0x00007ffff792717a in mfem::FiniteElementCollection::New (name=0x990950 "ND_R2D_2D_P2") at /home/rick/tmp/mfem/fem/fe_coll.cpp:335
#7  0x00007ffff793da17 in mfem::FiniteElementSpace::Load (this=this@entry=0x978ef0, m=m@entry=0x97c6d0, input=...) at /usr/include/c++/13/bits/basic_string.h:222
#8  0x00007ffff7967b97 in mfem::GridFunction::GridFunction (this=0x96f760, m=0x97c6d0, input=...) at /home/rick/tmp/mfem/fem/gridfunc.cpp:42
#9  0x00000000004221a5 in ReadSerial(StreamState&) ()
#10 0x000000000041f016 in main ()
```

This seems easy enough to fix, by adding a few lines to `fem/fe_coll.cpp`:

```cpp
...
else if (!strncmp(name, "ND_R2D",6))
{
      fec = new ND_R2D_FECollection(atoi(name+11),atoi(name + 7));
}
...
```

Now the correct finite element space gets loaded, but in exchange I get ugly segmentation faults. (the glvis window pops open and immediately crashes). The trace from gdb is given by:
```
#0  0x00007ffff5e195aa in _XEventsQueued () from /lib64/libX11.so.6
#1  0x00007ffff5df5e87 in XCheckIfEvent () from /lib64/libX11.so.6
#2  0x00007ffff6a0c3a5 in ?? () from /lib64/libSDL2-2.0.so.0
#3  0x00007ffff6951442 in ?? () from /lib64/libSDL2-2.0.so.0
#4  0x00007ffff6951867 in ?? () from /lib64/libSDL2-2.0.so.0
#5  0x000000000044a9ca in SdlMainThread::DispatchSDLEvents (this=this@entry=0x933e00 <GetMainThread()::inst>) at /home/rick/tmp/glvis/lib/sdl_main.cpp:191
#6  0x000000000044aea3 in SdlMainThread::MainLoop (this=0x933e00 <GetMainThread()::inst>, server=<optimized out>) at /home/rick/tmp/glvis/lib/sdl_main.cpp:140
#7  0x0000000000442ab4 in SdlWindow::StartSDL (server_mode=<optimized out>) at /home/rick/tmp/glvis/lib/sdl.cpp:85
#8  0x000000000042fb99 in SDLMainLoop (server_mode=<optimized out>) at /home/rick/tmp/glvis/lib/aux_vis.cpp:55
#9  0x000000000041f3d2 in main (argc=<optimized out>, argv=<optimized out>) at /home/rick/tmp/glvis/glvis.cpp:1400
```

I am not sure what these errors could mean, as I have no experience with SDL2. The suggested changes to `fem/fe_coll.cpp` seem to be necessary, but probably some changes are needed on glvis's side as well. Should I open a companion issue in the glvis issue tracker?

**Comment by Rickbude** [comment]
Although I didn't test this, I presume the visualization of `ND_R1D`, `RT_R1D` and `RT_R2D` will fail as well.

**Comment by mlstowell** [comment]
Hi, @Rickbude,

This is a good catch. I should have added those lines to `fe_coll.cpp`. You are also correct that changes will need to be made to GLVis. I've taken a quick look and I have two options for adding GLVis support. I'll discuss these options with the other MFEM developers before making any changes.

A companion issue in GLVis would be useful.

Best wishes,
Mark

**Comment by Rickbude** [comment]
Thanks for looking into this. I opened an identically-named issue in glvis: [https://github.com/GLVis/glvis/issues/291](url).

**Comment by mlstowell** [comment]
This has not been forgotten. It is tied to the development of a plasma physics mini app which is ramping up now.

**Comment by Rickbude** [comment]
Right, so before this issue goes stale again... I have made a branch `issue_4403` a while ago that adds the aforementioned lines to `fem/fe_coll.cpp` for `ND_R1D`, `ND_R2D`, `RT_R1D` and `RT_R2D`. (Are there more spaces that should be handled?)

With these changes, and after slightly modifying example 31 with tzanio's tip on the companion [glvis issue](https://github.com/GLVis/glvis/issues/291) like so:

```cpp
 // 2. Read the mesh from the given mesh file.  We can handle triangular,
 //    quadrilateral, or mixed meshes with the same code. In 1D and 2D, use
 //    SetCurvature to embed the mesh as surface mesh in 3D space.
 Mesh mesh(mesh_file, 1, 1);
 dim = mesh.Dimension();
 if (dim < 3)
 {
    mesh.SetCurvature(order,false,3);
 }
```

The aforementioned test doesn't crash anymore:

```
ex31 -m ../data/star.mesh -o 2
glvis -m refined.mesh -g sol.gf
```

I only tested the `ND_R2D`  space, but I assume the rest will work similarly. The reason for not making the branch into a PR yet, is threefold:

1) I am not 100% sure if `ND_R1D`, `ND_R2D`, `RT_R1D` and `RT_R2D` are the only finite element spaces to worry about.
2) I was thinking about adding tests, but I didn't get around to that.
3) I think the examples using these "special" space should be updated as well.

As far as I can see, only `ex31.cpp`, `ex31p.cpp` and `ex32p.cpp` use these "special" spaces. I could use `SetCurvature` in all of these examples, but for the parallel examples I am not sure if this should be done on the serial mesh, or on the parallel mesh. 

Furthermore, I am wondering if the examples should be modified even more. For example, the visualization part of example 31 spends a big portion of code on splitting the solution into xy-components and a z-component. For the solution this now seems a bit unnecessary, as we can now just visualize the ND_R2D solution directly. Unless this is for didactical purposes. For the curl, I think we can't really do better than the current code: I don't think the curl of an ND_R2D gridfunction is projected perfectly onto an RT_R2D gridfunction right? (that would need to be something like a RT x L2 space I guess).

**Comment by tzanio** [comment]
ping: @mlstowell

**Comment by mlstowell** [comment]
Hello, @Rickbude,

Thank you for tackling this.

Don't worry about reason number 1. Those four spaces would constitute a very useful addition. There are also some related trace spaces that come to mind but these don't need to be supported by GLVis.

Tests would be good but the PR reviewers can help identify the necessary tests.

Updating the examples is probably a good idea. For most 2D meshes the choice of applying SetCurvature in serial or parallel probably won't be significant. I would suggest doing it in parallel as a general rule but if that fails for some reason we can perform it in serial since the meshes are likely to have a small memory footprint.

I believe the RT_R2D space does contain the curl of the ND_R2D basis by design. If the surface were curved I might worry about this but even in that case it may just work. Regarding the visualization used in examples 31 and 32 I would suggest keeping the existing code at least temporarily for comparison. Another issue is VisIt/ParaView visualization. I know that VisIt in particular will probably require updates before it can visualize R2D fields using this or a similar trick. For this reason it would be good to keep the older visualization scheme to show VisIt users how to extract their results.

Best wishes,
Mark

**Comment by v-dobrev** [comment]
This issue is addressed in #4691. Please use that PR for further discussion.
